text/3668-async-closure.md

```diff
@@ -0,0 +1,541 @@
+
+Async closures act similarly to closures, and can have parts of their their signatures specified:
+
+```rust
+// They can have arguments annotated with types:
+let arg = async |x: i32| { async_add(x, 1).await };
+
+// They can have their return types annotated:
+let ret = async || -> Vec<Id> { async_iterator.collect().await };
+
+// They can be higher-ranked:
+let hr = async |x: &str| { do_something(x).await };
+```
+
+When called, they return an anonymous future type corresponding to the (not-yet-executed) body of the closure. These can be awaited like any other future.
+
+# Detailed Explanation
+
+### `AsyncFn*`
+
+This RFC introduces a family of `AsyncFn` traits. These traits are intended to remain perma-unstable to name or implement, just like the `Fn` traits. Nonetheless, we'll describe the details of these traits so as to explain the user-facing features enabled by them.
+
+The definition of the traits is (modulo `rustc_` attributes, and the `"rust-call"` ABI):
+
+
+> [!NOTE]
+> We omit some details about the `"rust-call"` calling convention and the fact that the `Args` parameter is enforced to be a tuple.
+
+```rust
+/// An async-aware version of the [`FnOnce`](crate::ops::FnOnce) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFnOnce<Args> {
+    /// Future returned by [`AsyncFnOnce::async_call_once`].
+    type CallOnceFuture: Future<Output = Self::Output>;
+
+    /// Output type of the called closure's future.
+    type Output;
+
+    /// Call the [`AsyncFnOnce`], returning a future which may move out of the called closure.
+    fn async_call_once(self, args: Args) -> Self::CallOnceFuture;
+}
+
+/// An async-aware version of the [`FnMut`](crate::ops::FnMut) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFnMut<Args>: AsyncFnOnce<Args> {
+    /// Future returned by [`AsyncFnMut::async_call_mut`] and [`AsyncFn::async_call`].
+    type CallRefFuture<'a>: Future<Output = Self::Output>
+    where
+        Self: 'a;
+
+    /// Call the [`AsyncFnMut`], returning a future which may borrow from the called closure.
+    fn async_call_mut(&mut self, args: Args) -> Self::CallRefFuture<'_>;
+}
+
+/// An async-aware version of the [`Fn`](crate::ops::Fn) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFn<Args>: AsyncFnMut<Args> {
+    /// Call the [`AsyncFn`], returning a future which may borrow from the called closure.
+    fn async_call(&self, args: Args) -> Self::CallRefFuture<'_>;
+}
+```
+
+### Associated types of `AsyncFn*` traits are not nameable
+
+Note that, unlike what is true today with the current `Fn*` traits, this RFC reserves as an implementation detail the associates types of the `AsyncFn*` traits, and these will not be nameable as part of the stable interface specified by this RFC.
```

Comment:
This makes it sound like I could tell from the declaration above that the associated types are no nameable ("note that" indicates that this is something I can notice by staring at the text before this note). I don't see how I could tell that though. Maybe use a less confusing wording.

Reply:
Sure. I don't really know what such a code example would look like, though.

The only difference here is that `FnOnce::Output` is nameable on stable today. This, however, was AFAICT only done as mitigation due to an unstable-ness checking bug: https://github.com/rust-lang/rust/pull/34365

So users cannot write `async fn foo<F: async Fn() -> U, U>(f: F) -> F::Output { f().await }`.

Reply:
Thanks.  We've added clarifications and a contrasting example.